xonsh/environ.py

```diff
@@ -786,228 +798,227 @@ def _yield_executables(directory, name):
-
-    return s
+def _get_parent_dir_for(path, dir_name, timeout):
+    # walk up the directory tree to see if we are inside an hg repo
+    # the timeout makes sure that we don't thrash the file system
+    previous_path = ''
+    t0 = time.time()
+    while path != previous_path and ((time.time() - t0) < timeout):
+        if os.path.isdir(os.path.join(path, dir_name)):
+            return path
+        previous_path = path
+        path, _ = os.path.split(path)
+    return (path == previous_path)
 
 
-@ensure_hg
 def get_hg_branch(cwd=None, root=None):
-    branch = None
-    active_bookmark = None
-
-    if root is not None:
-        branch_path = os.path.sep.join([root, '.hg', 'branch'])
-        bookmark_path = os.path.sep.join([root, '.hg', 'bookmarks.current'])
-
-        if os.path.exists(branch_path):
-            with open(branch_path, 'r') as branch_file:
-                branch = branch_file.read()
+    env = builtins.__xonsh_env__
+    cwd = env['PWD']
+    root = _get_parent_dir_for(cwd, '.hg', env['VC_BRANCH_TIMEOUT'])
+    if not isinstance(root, str):
+        # Bail if we are not in a repo or we timed out
+        if root:
+            return ''
         else:
-            branch = 'default'
+            return subprocess.TimeoutExpired(['hg'], env['VC_BRANCH_TIMEOUT'])
+    # get branch name
+    branch_path = os.path.sep.join([root, '.hg', 'branch'])
+    if os.path.exists(branch_path):
+        with open(branch_path, 'r') as branch_file:
+            branch = branch_file.read()
+    else:
+        branch = 'default'
+    # add bookmark, if we can
+    bookmark_path = os.path.sep.join([root, '.hg', 'bookmarks.current'])
+    if os.path.exists(bookmark_path):
+        with open(bookmark_path, 'r') as bookmark_file:
+            active_bookmark = bookmark_file.read()
+        branch = "{0}, {1}".format(*(b.strip(os.linesep) for b in
+                                   (branch, active_bookmark)))
+    else:
+        branch = branch.strip(os.linesep)
+    return branch
 
-        if os.path.exists(bookmark_path):
-            with open(bookmark_path, 'r') as bookmark_file:
-                active_bookmark = bookmark_file.read()
 
-    if active_bookmark is not None:
-        return "{0}, {1}".format(
-            *(b.strip(os.linesep) for b in (branch, active_bookmark)))
+_FIRST_BRANCH_TIMEOUT = True
 
-    return branch.strip(os.linesep) if branch else None
+def _first_branch_timeout_message():
+    global _FIRST_BRANCH_TIMEOUT
+    if not _FIRST_BRANCH_TIMEOUT:
+        return
+    _FIRST_BRANCH_TIMEOUT = False
+    print('xonsh: branch timeout: computing the branch name, color, or both '
+          'timed out while formatting the prompt. You may avoid this by '
+          'increaing the value of $VC_BRANCH_TIMEOUT or by removing branch '
+          'fields, like {curr_branch}, from your $PROMPT. See the FAQ '
+          'for more details. This message will be suppressed in the future.',
+          file=sys.stderr)
```

Comment:
Right now, it looks like this will be suppressed for the remainder of this session, but not globally.  Can we maybe add an option `$SUPPRESS_BRANCH_TIMEOUT_WARNING` (default `False`) and change the words a little, to say something like the following, maybe?

> ... This message will be suppressed for the remainder of this session.  To suppress this message permanently, set `$SUPPRESS_BRANCH_TIMEOUT_MESSAGE = True` in your `.xonshrc` file

and then only show it if `_FIRST_BRANCH_TIMEOUT` AND `$SUPPRESS_BRANCH_TIMEOUT_MESSAGE`?